core/src/main/java/io/opencensus/tags/Tag.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.tags;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Function;
+import io.opencensus.tags.TagKey.TagKeyBoolean;
+import io.opencensus.tags.TagKey.TagKeyLong;
+import io.opencensus.tags.TagKey.TagKeyString;
+import javax.annotation.concurrent.Immutable;
+
+/** {@link TagKey} paired with a value. */
+@Immutable
+public abstract class Tag {
```

Comment:
In my reference code (https://github.com/census-instrumentation/opencensus-java/pull/375/files?diff=split#diff-debe21dd6f7dd73fdfe20d37abb6b6d3), I parameterize the types for the TagKey and the TagValue, which allows us to have abstract getKey() and getValue() methods - it seems to me that both are of somewhat limited value at the base type (Tag) - thoughts?

The other direction would be to remove getKey() from the base type. Are there any advantages to having it here.